File: go/prompt.go
// prompt.go
//
// IPromptTemplateEngine + a fallback-only PromptTemplateEngine that
// renders the canonical Qwen/ChatML format.
//
// A full Jinja2-compatible renderer would pull in a non-stdlib dependency
// (pongo2 / gonja). For 1.5.0 parity we ship the fallback only; consumers
// that need catalog-driven custom chat_templates can wire their own
// IPromptTemplateEngine implementation.

package circleai

import (
	"strings"
)

// IPromptTemplateEngine renders chat messages into the prompt string the
// model expects.
type IPromptTemplateEngine interface {
	Render(modelDirectory string, messages []ChatMessage, addGenerationPrompt bool) string
}

// PromptTemplateEngine is the fallback-ChatML implementation.
// Returns the canonical Qwen/ChatML format. Catalog-driven custom
// chat_template strings (Jinja2 syntax) are out of scope for the Go
// port at this version.
type PromptTemplateEngine struct{}

// FallbackChatMLTemplate is the literal Jinja2 source the C# port
// falls back to; we don't render Jinja in Go, so this is exposed as a
// constant for documentation + cross-language fixture comparison.
const FallbackChatMLTemplate = `{%- for message in messages -%}
<|im_start|>{{ message.role }}
{{ message.content }}<|im_end|>
{% endfor -%}
{%- if add_generation_prompt -%}
<|im_start|>assistant
{%- endif -%}`

// Render implements IPromptTemplateEngine.
func (PromptTemplateEngine) Render(modelDirectory string, messages []ChatMessage, addGenerationPrompt bool) string {
	_ = modelDirectory // ignored — fallback always renders ChatML
	var b strings.Builder
	for _, m := range messages {
		role := normaliseRole(m.Role)
		b.WriteString("<|im_start|>")
		b.WriteString(role)
		b.WriteString("\n")
		b.WriteString(m.Content)
		b.WriteString("<|im_end|>\n")
	}
	if addGenerationPrompt {
		b.WriteString("<|im_start|>assistant\n")
	}
	return b.String()
}

func normaliseRole(role string) string {
	role = strings.ToLower(strings.TrimSpace(role))
	if role == "" {
		return "user"
	}
	if role == "tool" || role == "function" {
		return "user"
	}
	return role
}

File: rust/src/prompt.rs
//! prompt.rs
//!
//! Fallback ChatML prompt template engine. A real Jinja2-style renderer would
//! pull a runtime dep (minijinja / askama) which is out of scope for the
//! portable surface — host apps that need custom templates wire one in.

use crate::models::ChatMessage;

pub trait IPromptTemplateEngine: Send + Sync {
    fn render(
        &self,
        model_directory: &str,
        messages: &[ChatMessage],
        add_generation_prompt: bool,
    ) -> String;
}

pub const FALLBACK_CHAT_TEMPLATE: &str = r#"{%- for message in messages -%}
<|im_start|>{{ message.role }}
{{ message.content }}<|im_end|>
{% endfor -%}
{%- if add_generation_prompt -%}
<|im_start|>assistant
{%- endif -%}"#;

pub struct PromptTemplateEngine;

impl PromptTemplateEngine {
    pub fn new() -> Self {
        Self
    }
}

impl Default for PromptTemplateEngine {
    fn default() -> Self {
        Self::new()
    }
}

impl IPromptTemplateEngine for PromptTemplateEngine {
    fn render(
        &self,
        _model_directory: &str,
        messages: &[ChatMessage],
        add_generation_prompt: bool,
    ) -> String {
        let mut out = String::new();
        for m in messages {
            out.push_str("<|im_start|>");
            out.push_str(&normalise_role(&m.role));
            out.push('\n');
            out.push_str(&m.content);
            out.push_str("<|im_end|>\n");
        }
        if add_generation_prompt {
            out.push_str("<|im_start|>assistant\n");
        }
        out
    }
}

fn normalise_role(role: &str) -> String {
    let r = role.trim().to_ascii_lowercase();
    if r.is_empty() {
        return "user".to_string();
    }
    if r == "tool" || r == "function" {
        return "user".to_string();
    }
    r
}
